xapian-core/include/xapian/cluster.h

```diff
@@ -24,302 +24,392 @@
-    ClusterSetIterator operator++(int) {
-	ClusterSetIterator *c = this;
-	index++;
-	return *c;
-    }
+    /// This method is used to check the cluster at index 'i'
+    Cluster operator[](Xapian::doccount i);
 
-    // Prefix decrement operator for this iterator
-    ClusterSetIterator & operator--() {
-	index--;
-	return *this;
-    }
+    /// This method is used to clear all the Clusters in the ClusterSet
+    void clear_clusters();
 
-    // Postfix decrement operator for this iterator
-    ClusterSetIterator operator--(int) {
-	ClusterSetIterator c = *this;
-	index--;
-	return c;
-    }
+    /** This methood recalculates the centroids for all the centroids
+     *  in the ClusterSet
+     */
+    void recalculate_centroids();
 };
 
-// Equality test for ClusterSetIterator
-inline bool operator==(const ClusterSetIterator &a, const ClusterSetIterator &b) {
-    return (a.index == b.index);
-}
+/// This class represents an abstract class for a clusterer to be implemented
+class XAPIAN_VISIBILITY_DEFAULT Clusterer {
 
-// Inequality test for ClusterSetIterator
-inline bool operator!=(const ClusterSetIterator &a, const ClusterSetIterator &b) {
-    return !(a == b);
-}
+  public:
+
+    /// Destructor
+    virtual ~Clusterer();
 
-// Equality test for DocumentSetIterator
-inline bool operator==(const DocumentSetIterator &a, const DocumentSetIterator &b) {
-    return (a.index == b.index);
-}
+    /// This method helps implement the required clustering algorithm in the subclass
+    virtual ClusterSet cluster(MSet &mset) = 0;
 
-// Inequality test for DocumentSetIterator
-inline bool operator!=(const DocumentSetIterator &a, const DocumentSetIterator &b) {
-    return !(a == b);
-}
+    /// This method returns a description of the clusterer being used
+    virtual std::string get_description() = 0;
+};
 
 /** Round Robin clusterer:
  *  This clusterer is a minimal clusterer which will cluster documents as -
- *  ith document goes to the (i%N)th cluster where N is the number of clusters and
- *  0 <= i <= k; where k is the number of documents
+ *  ith document goes to the (i % k)th cluster where k is the number of clusters and
+ *  0 <= i < N; where N is the number of documents
  */
-class XAPIAN_VISIBILITY_DEFAULT RoundRobin {
+class XAPIAN_VISIBILITY_DEFAULT RoundRobin : public Clusterer {
+
+    /// This specifies the number of clusters to be formed by the clusterer
+    unsigned int num_of_clusters;
+
   public:
-    ~RoundRobin();
 
-    // Implements RoundRobin clustering
-    ClusterSet cluster(MSet &mset, unsigned int k);
+    /// Constructor
+    RoundRobin(unsigned int num_of_clusters_) : num_of_clusters(num_of_clusters_) {}
 
-    // Returns the description of the clusterer being used
+    /// This method implements the RoundRobin clustering
+    ClusterSet cluster(MSet &mset);
+
+    /// This method returns the description of the clusterer
     std::string get_description();
 };
-}
+};
```

Comment:
You don't need a `;` after `namespace { /*...*/ }` (`class X { /*...*/ }` needs one for irritating historical reasons - it's modelled on `struct X { /* ... */ };` in C).  I have a feeling it causes (or at least caused) a warning with one of the less common compilers.  I'm not sure if that's still the case (requiring C++11 has ruled out a few older compilers) but it's probably better avoided.  Certainly don't go around adding them, as you do here!